bpo-29576: Improve some deprecations in the importlib

Carreau opened the pull request:
Add the python version since the functionality is deprecated,
and raise a couple of deprecation warnings in a few places.

Theses functions are marked as deprecated in the documentation, but
especially in existing codebase, programmers tends to not re-check
whether functions are deprecated. So trigger the warning when possible.

It's also more probable that a developer will drop deprecated
functionality if we immediately give them information about
replacement API, and not have them to go find it in the documentation.

Include deprecation information in DocString as well, as many tools pull
documentation from there and not from docs.python.org.

berkerpeksag commented:
I fixed them now. Please note that this repository is a semi-official read-only mirror of hg.python.org/cpython. See https://docs.python.org/devguide/index.html for details.

CrazyPython commented:
@andportnoy Now that it has been updated, you should close this issue.

Carreau committed:
Improve some deprecations in the importlib

Add the python version since the functionality is deprecated,
and raise a couple of deprecation warnings in a few places.

Theses functions are marked as deprecated in the documentation, but
especially in existing codebase, programmers tends to not re-check
whether functions are deprecated. So trigger the warning when possible.

It's also more probable that a developer will drop deprecated
functionality if we immediately give them information about
replacement API, and not have them to go find it in the documentation.

Include deprecation information in DocString as well as many tools pull
documentation from there and not from docs.python.org.

Add test making sure `find_loader()` and `find_module()` Both emit a
deprecation warning.

fgregg commented on Lib/importlib/abc.py:
```diff
@@ -61,7 +64,13 @@ def find_module(self, fullname, path):
+        Deprecated since Python 3.4 in favor of find_spec (available since 3.4)
         """
+        warnings.warn("MetaPathFinder.find_module is deprecated since Python "
+                      "3.4 in favor of MetaPatFinder.find_spec "
```
typo here s/MetaPatFinder/MetaPathFinder/

Carreau replied:
Oops indeed, thanks.

brettcannon commented on Lib/importlib/abc.py:
```diff
@@ -97,7 +106,13 @@ def find_loader(self, fullname):
         This method is deprecated in favor of finder.find_spec(). If find_spec()
         is provided than backwards-compatible functionality is provided.
 
+        Deprecated since Python 3.4 in favor of find_spec (available since 3.4)
```
This line is somewhat redundant since the deprecation is mentioned in the previous paragraph. Maybe just add the version it was deprecated in above?

brettcannon commented on Lib/importlib/abc.py:
```diff
@@ -61,7 +64,13 @@ def find_module(self, fullname, path):
         exists then backwards-compatible functionality is provided for this
         method.
 
+        Deprecated since Python 3.4 in favor of find_spec (available since 3.4)
```
The deprecation is mentioned in the above paragraph, so if you would like to add the version info just toss in up there.

brettcannon commented on Lib/importlib/abc.py:
```diff
@@ -61,7 +64,13 @@ def find_module(self, fullname, path):
+        Deprecated since Python 3.4 in favor of find_spec (available since 3.4)
         """
+        warnings.warn("MetaPathFinder.find_module is deprecated since Python "
+                      "3.4 in favor of MetaPathFinder.find_spec "
```
Please add `()` after `find_spec`.

brettcannon commented on Lib/importlib/abc.py:
```diff
@@ -97,7 +106,13 @@ def find_loader(self, fullname):
+        Deprecated since Python 3.4 in favor of find_spec (available since 3.4)
         """
+        warnings.warn("PathEntryFinder.find_loader is deprecated since Python "
+                      "3.4 in favor of PathEntryFinder.find_spec "
```
Please add`()` as appropriate.

brettcannon reviewed (changes requested):
Could you add tests to make sure the warnings are being raised?

Carreau commented:
All comments should be addressed.

brettcannon commented:
Thanks! Please add entries in Misc/NEWS and one in What's New in the "deprecated" section and this will be ready to be merged!

Carreau committed:
Ad entry to Misc/NEWS and what's new 3.7

brettcannon commented:
Thanks!

Carreau commented:
Thanks !
